### PR TITLE
feat: Query connections feature

### LINKED ISSUE #429: Add ability to query all connections from did exchange service

### PR DESCRIPTION
**Pull request by sudeshrshetty**
- added Iterator() to store api interface to get iterator for given range.
- added query connections feature in connection recorder and client
- controller REST API to query all connections through client
- REST API to filter connections by state
- remaining query criteria will be applied as part of #655
- closes #429 

Signed-off-by: sudesh.shetty <sudesh.shetty@securekey.com>

### CODE REVIEW

File: pkg/client/didexchange/client.go [hunk 1]
```diff
@@ -153,12 +153,9 @@ func (c *Client) HandleInvitation(invitation *Invitation) error {
 }
 
 // QueryConnections queries connections matching given parameters
-func (c *Client) QueryConnections(request *QueryConnectionsParams) ([]*Connection, error) {
-	// TODO https://github.com/hyperledger/aries-framework-go/issues/429 query all connections from did exchange service
-	return []*Connection{
-		{&didexchange.ConnectionRecord{ConnectionID: uuid.New().String()}},
-		{&didexchange.ConnectionRecord{ConnectionID: uuid.New().String()}},
-	}, nil
+func (c *Client) QueryConnections(request *QueryConnectionsParams) ([]*didexchange.ConnectionRecord, error) {
```
Comment (troyronda) [inline pkg/client/didexchange/client.go]: TODO page results.

Comment (sudeshrshetty) [reply pkg/client/didexchange/client.go]: done

Comment (sudeshrshetty) [reply pkg/client/didexchange/client.go]: done

File: pkg/storage/store.go [hunk 2]
```diff
@@ -30,4 +30,42 @@ type Store interface {
...
+
+	// Key returns the key of the current key/value pair, or nil if done.
+	// The caller should not modify the contents of the returned slice, and
+	// its contents may change on the next call to any 'seeks method'.
+	Key() []byte
+
+	// Value returns the value of the current key/value pair, or nil if done.
+	// The caller should not modify the contents of the returned slice, and
+	// its contents may change on the next call to any 'seeks method'.
+	Value() []byte
```
Comment (soluchok) [inline pkg/storage/store.go]: I would rather create one `Result` function with a signature like this `Result() key,val,err` . In my opinion, this is more clear.

File: pkg/storage/leveldb/leveldb_store.go [hunk 3]
```diff
@@ -126,3 +128,11 @@ func (s *leveldbStore) Get(k string) ([]byte, error) {
 	}
 	return data, nil
 }
+
+// Iterator returns iterator for the latest snapshot of the underlying db.
+func (s *leveldbStore) Iterator(start, limit string) storage.StoreIterator {
+	if start == "" || limit == "" {
+		iterator.NewEmptyIterator(errors.New("start key is mandatory"))
```
Comment (soluchok) [inline pkg/storage/leveldb/leveldb_store.go]: wrong error message `start key is mandatory` better to change it `start or limit key is mandatory` or smth similar

File: pkg/storage/store.go (hunk 2 above)
Comment (soluchok) [reply pkg/storage/store.go]: oh, I see it is leveldb iterator

File: pkg/restapi/operation/didexchange/models/invitation.go [hunk 4]
```diff
@@ -9,6 +9,8 @@ package models
 import (
 	"time"
 
+	didexchangeSvc "github.com/hyperledger/aries-framework-go/pkg/didcomm/protocol/didexchange"
+
```
Comment (soluchok) [inline pkg/restapi/operation/didexchange/models/invitation.go]: extra line

Comment (sudeshrshetty) [reply pkg/restapi/operation/didexchange/models/invitation.go]: done

File: pkg/storage/leveldb/leveldb_store.go (hunk 3 above)
Comment (sudeshrshetty) [reply pkg/storage/leveldb/leveldb_store.go]: done

File: pkg/client/didexchange/client.go [hunk 5]
```diff
@@ -153,12 +153,10 @@ func (c *Client) HandleInvitation(invitation *Invitation) error {
...
```
Comment (rolsonquadras) [inline pkg/client/didexchange/client.go]: We shouldn't expose service pkg(didexchange.ConnectionRecord) to consumers/business logic(bdd/rest).

Comment (sudeshrshetty) [reply pkg/client/didexchange/client.go]: Yes I am fixing it in upcoming PR with bddtests. This one I pushed in hurry for demo. ALso, I see some swagger annotations used in client models which I am going to fix as well.

File: pkg/didcomm/protocol/didexchange/persistence.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ const (
 	myNSPrefix      = "my"
 	//Todo: Issue-556 It will not be constant, this namespace will need to be figured with verification key
 	theirNSPrefix = "their"
+	limitPattern  = "%s~"
```
Comment (troyronda) [inline pkg/didcomm/protocol/didexchange/persistence.go]: this constant could use explanation.

Comment (sudeshrshetty) [reply pkg/didcomm/protocol/didexchange/persistence.go]: will add it

Comment (sudeshrshetty) [reply pkg/didcomm/protocol/didexchange/persistence.go]: done

**Pull request merged**